include/cuco/detail/static_multimap/device_view_impl.inl

```diff
@@ -1055,6 +1055,216 @@ class static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view_
+   * @tparam ProbingCG Type of Cooperative Group used to retrieve
+   * @tparam OutputIt1 Device accessible output iterator whose `value_type` is constructible from
+   * `InputIt`s `value_type`.
+   * @tparam OutputIt2 Device accessible output iterator whose `value_type` is constructible from
+   * the map's `value_type`.
+   * @tparam PairEqual Binary callable type
+   * @param probing_cg The Cooperative Group used to retrieve
+   * @param pair The pair to search for
+   * @param probe_output_begin Beginning of the output sequence of the matched probe pairs
+   * @param contained_output_begin Beginning of the output sequence of the matched contained
+   * pairs
+   * @param pair_equal The binary callable used to compare two pairs for equality
+   */
+  template <bool is_outer,
+            bool uses_vector_load,
+            typename ProbingCG,
+            typename OutputIt1,
+            typename OutputIt2,
+            typename PairEqual>
+  __device__ __forceinline__ std::enable_if_t<uses_vector_load, void> pair_retrieve(
+    ProbingCG const& probing_cg,
+    value_type const& pair,
+    OutputIt1 probe_output_begin,
+    OutputIt2 contained_output_begin,
+    PairEqual pair_equal) noexcept
+  {
+    using cuda::std::memory_order_relaxed;
+
+    auto const lane_id                = probing_cg.thread_rank();
+    auto current_slot                 = initial_slot(probing_cg, pair.first);
+    [[maybe_unused]] auto found_match = false;
+
+    __shared__ atomic_ctr_type counter;
+    if (lane_id == 0) { counter.store(0, memory_order_relaxed); }
+
+    // no need to sync if `is_outer` is `true` due to the upcoming `cg.any()`
+    if constexpr (not is_outer) { probing_cg.sync(); }
+
+    while (true) {
+      value_type arr[2];
+      load_pair_array(&arr[0], current_slot);
+
+      auto const first_slot_is_empty =
+        detail::bitwise_compare(arr[0].first, this->get_empty_key_sentinel());
+      auto const second_slot_is_empty =
+        detail::bitwise_compare(arr[1].first, this->get_empty_key_sentinel());
+      auto const first_equals  = (not first_slot_is_empty and pair_equal(arr[0], pair));
+      auto const second_equals = (not second_slot_is_empty and pair_equal(arr[1], pair));
+
+      if constexpr (is_outer) {
+        auto const exists = probing_cg.any(first_equals or second_equals);
+        if (exists) { found_match = true; }
+      }
+
+      if (first_equals) {
+        auto output_idx = counter.fetch_add(1, memory_order_relaxed);
+        // TODO: `=` operator cannot work here
+        // *(probe_output_begin + output_idx)     = pair;
+        // *(contained_output_begin + output_idx) = arr[0];
+        thrust::get<0>(*(probe_output_begin + output_idx))     = pair.first;
+        thrust::get<1>(*(probe_output_begin + output_idx))     = pair.second;
+        thrust::get<0>(*(contained_output_begin + output_idx)) = arr[0].first;
+        thrust::get<1>(*(contained_output_begin + output_idx)) = arr[0].second;
```

Comment:
Why doesn't this work?

Reply:
no idea by what miracle the `=` operator works just fine right now.

Reply:
By writing unit tests for non-shmem `pair_retrieve`, I see the error again:
```
/home/yunsongw/Work/cuCollections/include/cuco/detail/static_multimap/device_view_impl.inl(1120): error: no operator "=" matches these operands
            operand types are: thrust::detail::tuple_of_iterator_references<thrust::detail::any_assign &, thrust::detail::any_assign &> = const cuco::pair<int32_t, int32_t>
```
and in this case, both `probe_output_begin` and `contained_output_begin` are defined as
```cpp
auto begin = thrust::make_zip_iterator(
    thrust::make_tuple(thrust::make_discard_iterator(), thrust::make_discard_iterator()));
```

Reply:
In light of our discussion offline, let's add a comment as to why we must assign the iterator elements individually rather than as a pair.